File: jarvis-personal/native/android/app/src/main/kotlin/com/dincr/app/AppModel.kt
package com.dincr.app

import android.app.Application
import android.content.Intent
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.dincr.data.ApiClient
import com.dincr.data.ApiError
import com.dincr.data.AppLanguage
import com.dincr.data.AuthException
import com.dincr.data.DincrService
import com.dincr.data.FixtureDincrService
import com.dincr.data.InMemorySessionStore
import com.dincr.data.LiveDincrService
import com.dincr.data.MoneyFormat
import com.dincr.data.OAuthProvider
import com.dincr.data.Pkce
import com.dincr.data.Profile
import com.dincr.data.SessionManager
import com.dincr.data.SupabaseAuthClient
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/** Build/launch configuration. No backend configured, or `dincrFixtures` extra → fixture data. */
sealed interface AppEnvironment {
    data class Live(val apiUrl: String, val supabaseUrl: String, val anonKey: String) : AppEnvironment
    data class Fixtures(val scenario: FixtureDincrService.Scenario, val skipLogin: Boolean, val latencyMs: Long = 350) : AppEnvironment

    companion object {
        const val AUTH_REDIRECT = "com.dincr.app://auth/callback"

        fun from(intent: Intent?): AppEnvironment {
            intent?.getStringExtra("dincrFixtures")?.let { raw ->
                val scenario = FixtureDincrService.Scenario.entries.firstOrNull { it.name.equals(raw, ignoreCase = true) }
                    ?: FixtureDincrService.Scenario.POPULATED
                // UI tests pass dincrLatencyMs=0: Compose's test dispatcher does not advance simulated network delays.
                return Fixtures(scenario, intent.getBooleanExtra("dincrSkipLogin", false), intent.getLongExtra("dincrLatencyMs", 350))
            }
            val api = BuildConfig.DINCR_API_URL
            val supabase = BuildConfig.DINCR_SUPABASE_URL
            val key = BuildConfig.DINCR_SUPABASE_ANON_KEY
            return if (api.isBlank() || supabase.isBlank() || key.isBlank()) Fixtures(FixtureDincrService.Scenario.POPULATED, false)
            else Live(api, supabase, key)
        }
    }
}

/** Gate order from CURRENT_STATE_AUDIT.md §1, same as the iOS AppModel. */
sealed interface Phase {
    data object Booting : Phase
    data object SignedOut : Phase
    data object LoadingIdentity : Phase
    data class IdentityError(val message: String) : Phase
    data object OwnerNotSupported : Phase
    data class NotYetSupported(val message: String) : Phase
    data object ProfileSetup : Phase
    data object Ready : Phase
}

class AppModel(application: Application) : AndroidViewModel(application) {
    private val _phase = MutableStateFlow<Phase>(Phase.Booting)
    val phase: StateFlow<Phase> = _phase.asStateFlow()
    private val _profile = MutableStateFlow<Profile?>(null)
    val profile: StateFlow<Profile?> = _profile.asStateFlow()
    private val _signInError = MutableStateFlow<String?>(null)
    val signInError: StateFlow<String?> = _signInError.asStateFlow()
    private val _signingIn = MutableStateFlow(false)
    val signingIn: StateFlow<Boolean> = _signingIn.asStateFlow()

    lateinit var environment: AppEnvironment private set
    lateinit var service: DincrService private set
    private var auth: SupabaseAuthClient? = null
    private lateinit var sessions: SessionManager
    private var pendingPkce: Pkce? = null
    private val language get() = AppLanguage.current()

    val isFixtures get() = environment is AppEnvironment.Fixtures
    val moneyFormat get() = MoneyFormat.from(_profile.value)

    /** Called once from the activity with its launch intent. */
    fun configure(environment: AppEnvironment) {
        if (this::environment.isInitialized) return
        this.environment = environment
        when (environment) {
            is AppEnvironment.Live -> {
                val client = SupabaseAuthClient(environment.supabaseUrl, environment.anonKey)
                auth = client
                sessions = SessionManager(client, KeystoreSessionStore(getApplication())) { viewModelScope.launch { signedOut() } }
                service = LiveDincrService(ApiClient(environment.apiUrl, sessions))
            }
            is AppEnvironment.Fixtures -> {
                sessions = SessionManager(null, InMemorySessionStore())
                service = FixtureDincrService(environment.scenario, environment.latencyMs)
            }
        }
        viewModelScope.launch {
            when {
                environment is AppEnvironment.Fixtures && environment.skipLogin -> loadIdentity()
                environment is AppEnvironment.Fixtures -> _phase.value = Phase.SignedOut
                sessions.hasSession -> loadIdentity()
                else -> _phase.value = Phase.SignedOut
            }
        }
    }

    /** URL for the Custom Tab, or null in fixture mode. */
    fun beginSignIn(provider: OAuthProvider): String? {
        _signInError.value = null
        val client = auth ?: return null
        return Pkce.generate().also { pendingPkce = it }.let { client.authorizeUrl(provider, AppEnvironment.AUTH_REDIRECT, it) }
    }

    fun handleCallback(uri: String) {
        val client = auth ?: return
        val pkce = pendingPkce ?: return
        pendingPkce = null
        viewModelScope.launch {
            _signingIn.value = true
            try {
                val code = SupabaseAuthClient.authorizationCode(uri, AppEnvironment.AUTH_REDIRECT)
                sessions.accept(client.exchange(code, pkce.verifier))
                loadIdentity()
            } catch (error: Exception) {
                _signInError.value = language.pick("No pudimos completar el acceso. Intentá nuevamente.", "We couldn’t sign you in. Please try again.")
            } finally {
                _signingIn.value = false
            }
        }
    }

    fun signInWithFixtures() = viewModelScope.launch {
        _signingIn.value = true
        loadIdentity()
        _signingIn.value = false
    }

    fun retryIdentity() = viewModelScope.launch { loadIdentity() }

    suspend fun loadIdentity() {
        if (_profile.value == null) _phase.value = Phase.LoadingIdentity
        try {
            apply(service.me())
        } catch (error: AuthException.SignedOut) {
            signedOut()
        } catch (error: ApiError) {
            _phase.value = Phase.IdentityError(error.message)
        }
    }

    fun apply(profile: Profile) {
        _profile.value = profile
        _phase.value = when {
            profile.isOwner -> Phase.OwnerNotSupported
            profile.legal?.required == true -> Phase.NotYetSupported(language.pick("Aceptá los términos actualizados desde la app actual de DINCR.", "Accept the updated terms in the current DINCR app."))
            profile.profileSetupCompleted != true -> Phase.ProfileSetup
            profile.planSelected == false -> Phase.NotYetSupported(language.pick("Elegí tu plan desde la app actual de DINCR.", "Choose your plan in the current DINCR app."))
            else -> Phase.Ready
        }
    }

    fun signOut() = viewModelScope.launch {
        sessions.signOut()
        signedOut()
    }

    private fun signedOut() {
        _profile.value = null
        _phase.value = Phase.SignedOut
    }
}

/** Short bilingual copy helper, same contract as the web `tx(es, en)` (prototype only). */
fun tx(spanish: String, english: String) = AppLanguage.current().pick(spanish, english)

File: jarvis-personal/native/ios/DINCR/App/AppModel.swift
import DincrCore
import Foundation
import Observation

/// Session and identity state. Reproduces the Capacitor gate order (CURRENT_STATE_AUDIT.md §1):
/// session → identity → Owner boundary → legal → profile setup → plan → app.
@MainActor @Observable
final class AppModel {
    enum Phase: Equatable {
        case booting
        case signedOut
        case loadingIdentity
        case identityError(String)
        case ownerNotSupported
        /// Legal consent and plan selection exist in the Capacitor app; the prototype routes
        /// those accounts to a notice until the modules land (PARITY_MATRIX A8, A11).
        case notYetSupported(String)
        case profileSetup
        case ready
    }

    private(set) var phase: Phase = .booting
    private(set) var profile: Profile?
    var signInError: String?
    private(set) var isSigningIn = false

    let service: DincrService
    let environment: AppEnvironment
    private let sessions: SessionManager
    private let auth: SupabaseAuthClient?
    private var pendingPKCE: PKCE?

    init(environment: AppEnvironment = .current()) {
        self.environment = environment
        switch environment.mode {
        case let .live(apiURL, supabaseURL, anonKey):
            let auth = SupabaseAuthClient(projectURL: supabaseURL, anonKey: anonKey)
            let sessions = SessionManager(auth: auth, store: KeychainSessionStore())
            self.auth = auth
            self.sessions = sessions
            self.service = LiveDincrService(client: APIClient(baseURL: apiURL, tokens: sessions))
        case let .fixtures(scenario):
            self.auth = nil
            self.sessions = SessionManager(auth: nil, store: InMemorySessionStore())
            self.service = FixtureDincrService(scenario: scenario)
        }
    }

    var language: AppLanguage { .current }
    var moneyFormat: MoneyFormat { MoneyFormat(profile: profile) }

    func start() async {
        await sessions.setSignedOutHandler { [weak self] in
            Task { @MainActor in self?.handleSignedOut() }
        }
        if environment.isFixtures {
            // Fixture sessions start signed out so the login screen is part of every flow.
            phase = ProcessInfo.processInfo.arguments.contains("-DincrSkipLogin") ? .loadingIdentity : .signedOut
            if phase == .loadingIdentity { await loadIdentity() }
            return
        }
        if await sessions.hasSession {
            await loadIdentity()
        } else {
            phase = .signedOut
        }
    }

    // MARK: Sign in

    /// The URL to open in the system authentication session.
    func beginSignIn(provider: OAuthProvider) -> URL? {
        signInError = nil
        guard let auth else { return nil }
        let pkce = PKCE.generate()
        pendingPKCE = pkce
        return auth.authorizeURL(provider: provider, redirectTo: AppEnvironment.authRedirect, pkce: pkce)
    }

    func completeSignIn(callback: URL) async {
        guard let auth, let pkce = pendingPKCE else { return }
        pendingPKCE = nil
        isSigningIn = true
        defer { isSigningIn = false }
        do {
            let code = try SupabaseAuthClient.authorizationCode(from: callback, redirect: AppEnvironment.authRedirect)
            let session = try await auth.exchange(code: code, verifier: pkce.verifier)
            await sessions.accept(session)
            await loadIdentity()
        } catch {
            signInError = language.pick("No pudimos completar el acceso. Intentá nuevamente.", "We couldn’t sign you in. Please try again.")
        }
    }

    func signInCancelled() {
        pendingPKCE = nil
    }

    /// Fixture mode: skip the provider and continue as the synthetic account.
    func signInWithFixtures() async {
        isSigningIn = true
        defer { isSigningIn = false }
        await loadIdentity()
    }

    // MARK: Identity

    func loadIdentity() async {
        if profile == nil { phase = .loadingIdentity }
        do {
            apply(try await service.me())
        } catch AuthError.signedOut {
            handleSignedOut()
        } catch let error as APIError {
            phase = .identityError(error.message)
        } catch {
            phase = .identityError(language.pick("No pudimos cargar tu cuenta.", "We couldn’t load your account."))
        }
    }

    func apply(_ profile: Profile) {
        self.profile = profile
        if profile.isOwner {
            phase = .ownerNotSupported
        } else if profile.legal?.required == true {
            phase = .notYetSupported(language.pick("Aceptá los términos actualizados desde la app actual de DINCR.", "Accept the updated terms in the current DINCR app."))
        } else if profile.profileSetupCompleted != true {
            phase = .profileSetup
        } else if profile.planSelected != true {
            phase = .notYetSupported(language.pick("Elegí tu plan desde la app actual de DINCR.", "Choose your plan in the current DINCR app."))
        } else {
            phase = .ready
        }
    }

    func signOut() async {
        await sessions.signOut()
        handleSignedOut()
    }

    private func handleSignedOut() {
        profile = nil
        phase = .signedOut
    }
}
